04-onion-routing.md

```diff
@@ -367,6 +379,155 @@ otherwise meets the amount criterion (eg. some other failure, or
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
+may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_data_tlv`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 6 (`path_id`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 8 (`next_blinding_override`)
+    2. data:
+        * [`point`:`blinding`]
+
+#### Requirements
+
+A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:
+
+- MUST create a blinded node ID `B(i)` for each node using the following algorithm:
+  - `e(0) <- {0;1}^256`
+  - `E(0) = e(0) * G`
+  - For every node `N(i)` in the route:
+    - let `P(i) = k(i) * G` be `N(i)`'s `node_id` (`k(i)` is `N(i)`'s private key)
+    - `ss(i) = SHA256(e(i) * P(i)) = SHA256(k(i) * E(i))` (ECDH shared secret known only by `N(r)` and `N(i)`)
+    - `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = SHA256(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
+- SHOULD create an `encrypted_data` for each blinded node `B(i)`
+- If it creates `encrypted_data`:
+  - MUST encrypt it with ChaCha20-Poly1305 using the `rho(i)` key and an all-zero nonce
+  - MAY store data in the `path_id` of the payload to itself to verify that
+    the route is used in the right context
+  - MAY add padding data to ensure all `encrypted_data` have the same length
+- MUST communicate the blinded node IDs `B(i)` and `encrypted_data(i)` to the sender
+- MUST communicate the real node ID of the introduction point `N(0)` to the sender
+- MUST communicate the first ephemeral key `E(0)` to the sender
+
+The sender:
+
+- MUST provide the `encrypted_data` field for each node in the blinded route
+  in their respective onion payloads
+- MUST communicate `E(0)` to the introduction point
+
+The introduction point:
+
+- MUST compute:
+  - `ss(0) = SHA256(k(0) * E(0))` (standard ECDH)
+  - `rho(0) = HMAC256("rho", ss(0))`
+  - `E(1) = SHA256(E(0) || ss(0)) * E(0)`
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(0)` and `E(0)`
+  - MUST use the decrypted fields to locate the next node
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
```

Comment:
```suggestion
- Otherwise:
```
Same typo again below. (I had a glimpe at this PR because @rustyrussell and I were talking about uses of AEAD in LN.)

Reply:
Thanks, fixed in https://github.com/lightning/bolts/pull/765/commits/d614959f7421f786c91b8332845c1fcd29499069.
If you want to have more than a glimpse, I'd be happy to get a review on the crypto ;)